bsp/src/org/jetbrains/bsp/project/test/bspTestRunConfig.scala

```diff
@@ -28,17 +31,61 @@ class BspTestRunFactory(t: ConfigurationType) extends ConfigurationFactory(t) {
   override def getName: String = "BspTestRunFactory"
 }
 
+class BspTestSettingsEditor(val proj: Project) extends SettingsEditor[BspTestRunConfiguration] {
+
+  val form = new BspTestConfigurationForm(proj)
+
+  override def resetEditorFrom(settings: BspTestRunConfiguration): Unit = form(settings)
+
+  override def applyEditorTo(settings: BspTestRunConfiguration): Unit = settings(form)
+
+  override def createEditor(): JComponent = form.mainPanel
+}
+
+
+class BspNoMatchingClassException() extends ConfigurationException(
+  "No class is matching with the regex",
+  s"The test class  was not reported by BSP endpoint buildTarget/scalaTestClasses (try to reimport the project)")
+
 
 class BspTestRunConfiguration(val project: Project, val configurationFactory: ConfigurationFactory, val name: String)
   extends RunConfigurationBase[String](project, configurationFactory, name) {
-  override def getConfigurationEditor: SettingsEditor[_ <: RunConfiguration] = new SettingsEditor[RunConfiguration]() {
-    override def resetEditorFrom(s: RunConfiguration): Unit = {}
 
-    override def applyEditorTo(s: RunConfiguration): Unit = {}
+  var testMode: TestMode = TestMode.ALL_IN_PROJECT
+  var matchedTestClasses: Map[URI, List[String]] = Map.empty
+  var testClassesRegex = ".*"
 
-    //TODO create editor
-    override def createEditor(): JComponent = new JComponent {}
+  def apply(settings: BspTestConfigurationForm): Unit = {
+    testMode = settings.getTestMode
+    testClassesRegex = settings.getTestClassNameRegex
+    matchedTestClasses = testMode match {
+      case TestMode.CLASS =>
+        val testClassRegex = try {
```

Comment:
This performs almost the same logic as `BspTestconfigurationForm.onClassNameRegexChanged()`. This is called on even more frequently, since every UI change in the run config editor tries to UI the data to a run configuration. I have tested for several hundred test classes, the UI was smooth, but still, I think it can be done better.